consensus/src/test/kotlin/maru/consensus/qbft/MinimalQbftMessageDecoderTest.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Consensys Software Inc.
+ *
+ * This file is dual-licensed under either the MIT license or Apache License 2.0.
+ * See the LICENSE-MIT and LICENSE-APACHE files in the repository root for details.
+ *
+ * SPDX-License-Identifier: MIT OR Apache-2.0
+ */
+package maru.consensus.qbft
+
+import java.util.Optional
+import kotlin.random.Random
+import maru.consensus.qbft.adapters.QbftBlockAdapter
+import maru.consensus.qbft.adapters.QbftBlockCodecAdapter
+import maru.core.ext.DataGenerators
+import org.apache.tuweni.bytes.Bytes
+import org.assertj.core.api.Assertions.assertThat
+import org.hyperledger.besu.consensus.common.bft.ConsensusRoundIdentifier
+import org.hyperledger.besu.consensus.common.bft.payload.SignedData
+import org.hyperledger.besu.consensus.qbft.core.messagedata.CommitMessageData
+import org.hyperledger.besu.consensus.qbft.core.messagedata.PrepareMessageData
+import org.hyperledger.besu.consensus.qbft.core.messagedata.ProposalMessageData
+import org.hyperledger.besu.consensus.qbft.core.messagedata.QbftV1
+import org.hyperledger.besu.consensus.qbft.core.messagedata.RoundChangeMessageData
+import org.hyperledger.besu.consensus.qbft.core.messagewrappers.Commit
+import org.hyperledger.besu.consensus.qbft.core.messagewrappers.Prepare
+import org.hyperledger.besu.consensus.qbft.core.messagewrappers.Proposal
+import org.hyperledger.besu.consensus.qbft.core.messagewrappers.RoundChange
+import org.hyperledger.besu.consensus.qbft.core.payload.CommitPayload
+import org.hyperledger.besu.consensus.qbft.core.payload.PreparePayload
+import org.hyperledger.besu.consensus.qbft.core.payload.ProposalPayload
+import org.hyperledger.besu.consensus.qbft.core.payload.RoundChangePayload
+import org.hyperledger.besu.consensus.qbft.core.types.QbftMessage
+import org.hyperledger.besu.crypto.SignatureAlgorithmFactory
+import org.hyperledger.besu.datatypes.Hash
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.MessageData
+import org.junit.jupiter.api.Test
+
+class MinimalQbftMessageDecoderTest {
+  private val signatureAlgorithm = SignatureAlgorithmFactory.getInstance()
+  private val keyPair = signatureAlgorithm.generateKeyPair()
+  private val sequenceNumber = 100
+  private val roundNumber = 15
+  private val roundIdentifier = ConsensusRoundIdentifier(sequenceNumber.toLong(), roundNumber)
+  private val blockHash = Hash.hash(Bytes.wrap(Random.nextBytes(32)))
+
+  @Test
+  fun `should decode Prepare message`() {
+    val preparePayload = PreparePayload(roundIdentifier, blockHash)
+    val signature = signatureAlgorithm.sign(preparePayload.hashForSignature(), keyPair)
+    val signedPayload = SignedData.create(preparePayload, signature)
+    val prepare = Prepare(signedPayload)
+    val messageData = PrepareMessageData.create(prepare)
+    val qbftMessage = TestQbftMessage(messageData)
+
+    val metadata = MinimalQbftMessageDecoder.deserialize(qbftMessage)
```

Comment:
`signatureToAddress` inside `deserialize` is relying on a non-injectable `org.hyperledger.besu.ethereum.core.Util.signatureToAddress`. I think it's a big testability and design flaw, because: 
1. There is more than 1 place in the code to change the signature algorithm
2. We have to use the real signature algorithm and we have to copy-paste the way to use it from the dependency's source code (`SignatureAlgorithmFactory.getInstance()`)

Reply:
I've made the existing `Crypto` object an interface now and moved this current implementation into `SecpCrypto`. And have change the `MinimalQbftMessageDecoder` to use this.

I haven't changed all the existing uses of `Crypto` to inject the implementation so am just using `SecpCrypto` for now. I think this should be done in another PR. It's reasonably simple but will change a lot of files.